rca/utils/formatters.py

```diff
@@ -0,0 +1,96 @@
+    "ResearchCentrePage": "RESEARCH CENTRE",
+    "ShortCoursePage": "SHORT COURSE",
+    "ProgrammePage": "PROGRAMME",
+}
+
+
+def related_list_block_slideshow(slides):
+    # This formatter can be used to render out slides when defined on the model
+    # with RelatedPageListBlockPage()
+    # E.G slides = StreamField(StreamBlock([("Page", RelatedPageListBlockPage())]))
+    # The reason being that custom field data OR internal pages can be referenced
+    formated_slides = []
+    for slide in slides:
+        for block in slide.value:
+            if block.block_type == "custom_teaser":
+                formated_slides.append(
+                    {
+                        "value": {
+                            "title": block.value["title"],
+                            "summary": block.value["text"],
+                            "image": block.value["image"],
+                            "link": block.value["link"]["url"],
+                            "type": block.value["meta"],
+                        }
+                    }
+                )
+            elif block.block_type == "page":
+                page_type = None
+                page = block.value.specific
+                if page.__class__.__name__ in page_type_mapping:
+                    page_type = page_type_mapping.get(page.__class__.__name__, None)
+                summary = (
+                    page.introduction
+                    if hasattr(page, "introduction")
+                    else page.listing_summary
+                )
+                image = (
+                    page.hero_image
+                    if hasattr(page, "hero_image")
+                    else page.listing_image
+                )
+                formated_slides.append(
+                    {
+                        "value": {
+                            "title": page.title,
+                            "summary": summary,
+                            "image": image,
+                            "link": page.url,
+                            "type": page_type,
+                        }
+                    }
+                )
+    return formated_slides
+
+
+def format_page_teasers(obj):
+    if not obj:
+        return
+    page_teasers = {"title": obj.title, "summary": obj.summary, "pages": []}
+    for item in obj.pages:
+        for block in item.value:
+            if block.block_type == "custom_teaser":
+                page_teasers["pages"].append(
+                    {
+                        "title": block.value["title"],
+                        "description": block.value["text"],
+                        "image": block.value["image"],
+                        "link": block.value["link"]["url"],
+                        "type": block.value["meta"],
+                    }
+                )
+            elif block.block_type == "page":
+                page = block.value.specific
+                summary = (
+                    page.introduction
```

Comment:
I think you can replace `if hasattr() else` with `getattr(page, 'introduction', page.listing_summary)`, however neither will show `page.listing_summary` if `page.introduction` exists but is empty (assuming its optional)

Reply:
Yea this one is a bit strange. Seen as though I've made this global and re-usable now I've changed it to actually fall back to the listing summary. E.G https://github.com/torchbox/rca-wagtail-2019/pull/510/files#diff-9986ad6af90b2b99d634e41d81c2324a220e575c7afb1cbc546c30a1fcc8b792R33

```[python]
                summary = (
                    page.introduction
                    if hasattr(page, "introduction") and page.introduction
                    else page.listing_summary
                )
```